components/purchase_modal/purchase_modal.tsx

```diff
@@ -36,54 +39,69 @@ let stripePromise: Promise<Stripe | null>;
 type Props = {
     show: boolean;
     isDevMode: boolean;
-    products?: Dictionary<Product>;
+    products: Dictionary<Product> | undefined;
     contactSupportLink: string;
     contactSalesLink: string;
     isFreeTrial: boolean;
+    productId: string | undefined;
     actions: {
         closeModal: () => void;
         getCloudProducts: () => void;
         completeStripeAddPaymentMethod: (stripe: Stripe, billingDetails: BillingDetails, isDevMode: boolean) => Promise<boolean | null>;
+        subscribeCloudSubscription: (productId: string) => Promise<boolean | null>;
         getClientConfig: () => void;
         getCloudSubscription: () => void;
     };
 }
 
 type State = {
     paymentInfoIsValid: boolean;
-    productPrice: number;
     billingDetails: BillingDetails | null;
     cardInputComplete: boolean;
     processing: boolean;
+    selectedProduct: Product | null | undefined;
+}
+
+function findProductInDictionary(products: Dictionary<Product> | undefined, productId?: string): Product | null {
+    if (!products) {
+        return null;
+    }
+    const keys = Object.keys(products);
+    if (!keys.length) {
+        return null;
+    }
+    if (!productId) {
+        return products[keys[0]];
+    }
+    let selectedProduct = null;
+    if (keys.length > 1) {
+        // here find the product by the provided id, otherwise return the one with Professional in the name
```

Comment:
Given that we already check for `key.length` above this can probably be removed as well and to simplify even further we can just initialized with the first one and replace in the loop if found. 
```suggestion
    let selectedProduct = products[keys[0]];;
    if (keys.length > 1) {
        // here find the product by the provided id, otherwise return the one with Professional in the name
```